### PR TITLE
PP-12444 simplified account settings responsible person stripe task

### PR DESCRIPTION
**Pull request by nlsteers**
## WHAT

- controllers, views and validation functions for new multi-part responsible person form
  - users now see a check your answers page before submitting the responsible person details
- new checkTaskCompletion middleware and accompanying error class to prevent users from visiting task pages they have already completed
- new nunjucks custom filter to display dates in standard format summary tables etc

### SCREENS

Screenshot 2024-11-12 at 11 27 46

Screenshot 2024-11-12 at 11 27 54

Screenshot 2024-11-12 at 11 28 28

Screenshot 2024-11-12 at 11 28 42

Screenshot 2024-11-12 at 11 29 02

Screenshot 2024-11-12 at 11 29 25

Screenshot 2024-11-12 at 11 29 42

Screenshot 2024-11-12 at 11 30 04

### CODE REVIEW

File: app/controllers/simplified-account/settings/stripe-details/responsible-person/responsible-person-check-answers.controller.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    email: contact.workEmail,
+    address_line1: address.homeAddressLine1,
+    address_city: address.homeAddressCity,
+    address_postcode: address.homeAddressPostcode,
+    ...(address.homeAddressLine2 && { address_line2: address.homeAddressLine2 })
+  }
+  try {
+    await updateStipeDetailsResponsiblePerson(req.service, req.account, responsiblePerson)
+    delete req.session.formData
+    res.redirect(formatSimplifiedAccountPathsFor(paths.simplifiedAccount.settings.stripeDetails.index, req.service.externalId, req.account.type))
```
Comment (oswaldquek) [inline app/controllers/simplified-account/settings/stripe-details/responsible-person/responsible-person-check-answers.controller.js:44]: I'd prefer lines 43 and 44 to be outside of the try block. This would make it clear errors can only be catched from `updateStipeDetailsResponsiblePerson`

File: app/utils/simplified-account/validation/responsible-person.schema.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  dob: {
+    validate: body('dobDay') // Use just one DOB field as the anchor to prevent multiple validations
+      .custom((value, { req }) => {
+        const { dobDay, dobMonth, dobYear } = req.body
+        if (!dobDay && !dobMonth && !dobYear) {
+          throw new Error('Enter the date of birth')
+        }
+
+        // Stripe requires minimum age of 13
+        const day = Number(dobDay)
```
Comment (oswaldquek) [inline app/utils/simplified-account/validation/responsible-person.schema.js:32]: Does the [dobDay validation](https://github.com/alphagov/pay-selfservice/pull/4323/files#diff-2bd10cbbf12a432d034d00530df64d47e0663f73ad0032e46eb5226ae4ad145dR52) get executed before this line gets executed?

File: app/utils/simplified-account/validation/responsible-person.schema.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        const month = Number(dobMonth) - 1
+        const year = Number(dobYear)
+
+        const dob = new Date(year, month, day)
```
Comment (oswaldquek) [inline app/utils/simplified-account/validation/responsible-person.schema.js:36]: Months seems to be zero-based in Node 😬 Furthermore dates roll over if an invalid day of month is given 😬 
```> new Date(1999, 02, 31)
1999-03-30T23:00:00.000Z
> new Date(1999, 01, 31)
1999-03-03T00:00:00.000Z
> new Date(1999, 01, 01)
1999-02-01T00:00:00.000Z
> new Date(1999, 0, 01)
1999-01-01T00:00:00.000Z
> new Date(1999, 01, 32)
1999-03-04T00:00:00.000Z
```
I think we need a test for `responsible-person.schema.js`. But happy for this to go in a different PR.

File: app/simplified-account-routes.js [hunk 4]
```diff
@@ -39,7 +39,16 @@ stripeDetailsRouter.post(stripeDetailsPath.bankAccount, serviceSettingsControlle
...
+stripeDetailsRouter.post(stripeDetailsPath.responsiblePerson.index, serviceSettingsController.stripeDetails.responsiblePerson.post)
+stripeDetailsRouter.get(stripeDetailsPath.responsiblePerson.homeAddress, serviceSettingsController.stripeDetails.responsiblePerson.homeAddress.get)
+stripeDetailsRouter.post(stripeDetailsPath.responsiblePerson.homeAddress, serviceSettingsController.stripeDetails.responsiblePerson.homeAddress.post)
+stripeDetailsRouter.get(stripeDetailsPath.responsiblePerson.contactDetails, serviceSettingsController.stripeDetails.responsiblePerson.contactDetails.get)
+stripeDetailsRouter.post(stripeDetailsPath.responsiblePerson.contactDetails, serviceSettingsController.stripeDetails.responsiblePerson.contactDetails.post)
+stripeDetailsRouter.get(stripeDetailsPath.responsiblePerson.checkYourAnswers, serviceSettingsController.stripeDetails.responsiblePerson.checkYourAnswers.get)
+stripeDetailsRouter.post(stripeDetailsPath.responsiblePerson.checkYourAnswers, serviceSettingsController.stripeDetails.responsiblePerson.checkYourAnswers.post)
+// --
```
Comment (oswaldquek) [inline app/simplified-account-routes.js:51]: Do we not need any permissions for these?

File: app/utils/custom-nunjucks-filters/govuk-date.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+const { DateTime } = require('luxon')
```
Comment (oswaldquek) [inline app/utils/custom-nunjucks-filters/govuk-date.js:1]: Where is this file being used?

File: app/utils/simplified-account/validation/responsible-person.schema.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        const today = new Date()
+
+        let age = today.getFullYear() - dob.getFullYear()
+        const monthDiff = today.getMonth() - dob.getMonth()
+
+        if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < dob.getDate())) {
+          age--
+        }
```
Comment (oswaldquek) [inline app/utils/simplified-account/validation/responsible-person.schema.js:44]: Wish there is a nicer API around this 😬

File: app/utils/simplified-account/validation/responsible-person.schema.js (hunk 3 above)
Comment (nlsteers) [reply app/utils/simplified-account/validation/responsible-person.schema.js:36]: yeah I minus the month on line 33

File: app/controllers/simplified-account/settings/stripe-details/bank-account/bank-account.controller.test.js [hunk 7]
```diff
@@ -41,7 +43,7 @@ const setupTest = (method, additionalStubs = {}, additionalResProps = {}, additi
     ...additionalReqProps
   }
   next = sinon.spy()
-  bankAccountController[method](req, res, next)
+  bankAccountController[method][1](req, res, next)
```
[the remark this answers is not in the document]
Comment (nlsteers) [reply app/controllers/simplified-account/settings/stripe-details/bank-account/bank-account.controller.test.js:46]: the controller export is actually an array now:

```js
module.exports = {
  get: [checkTaskCompletion(stripeDetailsTasks.bankAccount.name), get],
  post: [checkTaskCompletion(stripeDetailsTasks.bankAccount.name), post]
}
```
this test is concerned with the get function, not the middleware so we're just calling that

File: app/controllers/simplified-account/settings/stripe-details/responsible-person/responsible-person-check-answers.controller.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  } catch (err) {
+    if (err && err.type === 'StripeInvalidRequestError') {
```
[the remark this answers is not in the document]
Comment (nlsteers) [reply app/controllers/simplified-account/settings/stripe-details/responsible-person/responsible-person-check-answers.controller.js:46]: yes this is likely overly defensive

File: app/simplified-account-routes.js (hunk 4 above)
Comment (nlsteers) [reply app/simplified-account-routes.js:51]: see line 34/35

the stripeDetailsRouter has the permission middleware, this way all routes that use that router have the same permission without constantly repeating ourselves

File: app/utils/simplified-account/validation/responsible-person.schema.js (hunk 6 above)
Comment (nlsteers) [reply app/utils/simplified-account/validation/responsible-person.schema.js:44]: yeah it's not pretty

File: app/utils/simplified-account/validation/responsible-person.schema.js (hunk 2 above)
Comment (nlsteers) [reply app/utils/simplified-account/validation/responsible-person.schema.js:32]: no, they are run in sequence with some conditions on the individual inputs to only run if some values are already set

File: app/utils/custom-nunjucks-filters/govuk-date.js (hunk 5 above)
Comment (nlsteers) [reply app/utils/custom-nunjucks-filters/govuk-date.js:1]: see https://github.com/alphagov/pay-selfservice/pull/4323/files#diff-6c6729ab30442e4e21ba0ca2f077494c1b57e7814c23ea9f9e4e50beac36149eR48

File: app/simplified-account-routes.js (hunk 4 above)
Comment (oswaldquek) [reply app/simplified-account-routes.js:51]: Oh, that's nice.

File: app/utils/simplified-account/validation/responsible-person.schema.js (hunk 3 above)
Comment (oswaldquek) [reply app/utils/simplified-account/validation/responsible-person.schema.js:36]: Do you plan to do further testing around the validations?

Comment (nlsteers) [reply app/utils/simplified-account/validation/responsible-person.schema.js:36]: yes, this PR is quite large. I have cypress and controller tests coming shortly

**Review by oswaldquek — APPROVED** [review]

**Review by oswaldquek — APPROVED** [review]
